AudiusProject/audius-protocol
# ⚠️  Upload Saga Typed and Parallel Stems v2

# Description

- Rewrites the upload saga, this time in typescript
- Fixes a bunch of type issues in the upload form
- Consolidates duplicate types
- Adds copious amounts of notes for future explorers
- Parallel stems are now associated with their parent throughout the whole process
- Progress for stems is nested with their parent track
- Simplified a lot of code

Did my best to not touch too much functionality and stick to types where possible but I couldn't help myself in some places.

TODO:
  - [x] Tests for progress selectors
  - [x] Test for handleUpload for tracks
  - [ ] Test for handleUpload for collection
  - [ ] Test for handleUpload for stems
  - [x] Test for handleUpload for tracks with stems
  - [x] Fix cypress integration tests
- [ ] Explore adding back old saga and using feature flag to switch which one is running to mitigate risk

Uploaded a79a74bccff2c2023a510b9ed3964b55255ef9ea to my subdomain, see if you can break it.

# Testing

Things touched
- Single track uploads
- Multi track uploads
- Album uploads
- Playlist uploads
- Track edit (especially stems)
- All the above uploads with and without stems
- Error handling

Subtle things to test potentially:
- USDC Gated prices and purchasing of uploaded tracks works
- Rights/Attribution works
- Remixes work
- Navigating away during upload
- Upload timeouts (I got rid of confirmer, so things won't time out anymore at least from confirmer - requests should still time out as normal though?)
- Analytics parity (I changed it so that analytics fire immediately rather than waiting til the end) (especially check the stem upload in edit track modal matches analytics)
- No duplicate cover art uploads for tracks within collections
- Do tracks uploaded in playlists need a metadata field for it? `is_playlist_upload: isCollection` is this still used?
- Make sure even if track upload succeeds, all tracks show like expected!
- Adding stems to existing tracks
- Editing tracks
- Editing collections
